### ISSUE TITLE
How to handle missing/duplicate field names in Spanner?

### ISSUE THREAD

**Issue by blowmage** [GoogleCloudPlatform/google-cloud-ruby#1449]
In spanner the returned data can have missing/empty field names.

```ruby
db.execute("SELECT 1").rows.to_a #=> [{:""=>1}]
```

The first question is how to handle multiple missing/empty field names. The results are returned in Ruby hashes, and Hash does not allow duplicate field names, so the empty string symbol (`:""`) used for both columns is overridden.

```ruby
db.execute("SELECT 1, 2").rows.to_a #=> [{:""=>2}]
```

The second question is a variation on the first. How do we handle duplicate field names. For example:

```ruby
db.execute("SELECT 1 AS count, 2 AS count").rows.to_a #=> [{:count=>2}]
```

/cc @vkedia

**Comment by blowmage** [comment]
My first impulse to the first question is to replace missing/empty field names with a default name that includes the column position. For example:

```ruby
db.execute("SELECT 1, 2").rows.to_a #=> [{:_col_0=>1, :_col_1=>2}]
```

But, this doesn't address the second question, since only the empty field names would be changed.

One options is to find all duplicates, and append a count to them:

```ruby
db.execute("SELECT 1, 2, 3 AS count, 4 AS count").rows.to_a
#=> [{:""=>2, :_1=>2, :count=>3, :count_1=>4}]
```

A variation on this is to modify all filed names that have a duplicate, not just the duplicate fields:

```ruby
db.execute("SELECT 1, 2, 3 AS count, 4 AS count").rows.to_a
#=> [{:_0=>2, :_1=>2, :count_0=>3, :count_1=>4}]
```

The returned field names would also be modified, so the names would match up. Currently, we have:

```ruby
db.execute("SELECT 1, 'hello'").types #=> {:""=>:STRING}
```

So we need some way to differentiate the field names.

```ruby
db.execute("SELECT 1, 'hello'").types #=> {:_0=>:INT64, :_1=>:STRING}
```

**Comment by blowmage** [comment]
One possible solution is return a Hash that has values by name _and_ position. The duplicate names would be squashed just like it is now:

````ruby
db.execute("SELECT 1, 2, 3, 4").rows.to_a
#=> [{:""=>4, 0=>1, 1=>2, 2=>3, 3=>4}]

db.execute("SELECT 1 AS count, 2 AS count, 3 AS count, 4 AS count").rows.to_a
#=> [{:count=>4, 0=>1, 1=>2, 2=>3, 3=>4}]
````

We can also add additional `type` methods to allow retrieving both the type names and type values as arrays:

````ruby
results = db.execute("SELECT 1, 2, 3, 4")
results.types #=> {:""=>: INT64}
results.type_names #=> [:"", :"", :"", :""]
results.type_values #=> [:INT64, :INT64, :INT64, :INT64]

results.rows.to_a #=> [{:"=>4, 0=>1, 1=>2, 2=>3, 3=>4}]
````

**Comment by blowmage** [comment]
Yet another possible solution is to add a flag to control whether the results are returned as hashes or a deeply nested arrays.

```ruby
hash_results = db.execute "SELECT 1, 2, 3, 4"
hash_results.types #=> {:""=>: INT64}
hash_results.rows.to_a #=> [{:""=>4}]

# Make a second query since #rows enumerator can only be called once
array_results = db.execute "SELECT 1, 2, 3, 4"
array_results.types array: true #=> [[:""=> :INT64], [:""=> :INT64], [:""=> :INT64], [:""=> :INT64]]
array_results.rows(array: true).to_a #=> [[[:""=>1], [:""=>2], [:""=>3], [:""=>4]]]
```

**Comment by blowmage** [comment]
That last option is exactly what the current code is doing, minus the last step that converts the deeply nested array of pairs into a Hash. The row as an array of pairs looks like this:

```ruby
row = [[:"", 1], [:"", 2], [:"", 3], [:"", 4]]
```

And when we convert that row of pairs into an Hash we get this:

```ruby
row = [[:"", 1], [:"", 2], [:"", 3], [:"", 4]]
Hash[row] #=> {:""=>4}
```

**Comment by blowmage** [comment]
This was resolved in #1461.
